src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/NoOpRestoreUtilities.cs

```diff
@@ -226,12 +226,68 @@ public static string GetHash(RestoreRequest request)
                 {
                     projectSpec.RestoreSettings = null;
                 }
+                PersistHashedDGFileIfDebugging(dgSpec, request.Log);
                 return dgSpec.GetHash();
             }
 
+            PersistHashedDGFileIfDebugging(request.DependencyGraphSpec, request.Log);
             return request.DependencyGraphSpec.GetHash();
         }
 
+
+        /// <summary>
+        /// Write the dg file to a temp location if NUGET_PERSIST_NOOP_DG.
+        /// </summary>
+        /// <remarks>This is a noop if NUGET_PERSIST_NOOP_DG is not set to true.</remarks>
+        private static void PersistHashedDGFileIfDebugging(DependencyGraphSpec spec, ILogger log)
+        {
+            if (_isPersistDGSet.Value)
+            {
+                string path;
+                var envPath = Environment.GetEnvironmentVariable("NUGET_PERSIST_NOOP_DG_PATH");
+                if (!string.IsNullOrEmpty(envPath))
+                {
+                    path = envPath;
+                    Directory.CreateDirectory(Path.GetDirectoryName(path));
+                }
+                else
+                {
+                    path = Path.Combine(
+                        NuGetEnvironment.GetFolderPath(NuGetFolderPath.Temp),
+                        "nuget-dg",
+                        $"{spec.Projects.FirstOrDefault().RestoreMetadata.ProjectName}-{DateTime.Now.ToString("yyyyMMddHHmmss")}.dg");
+                    DirectoryUtility.CreateSharedDirectory(Path.GetDirectoryName(path));
+                }
+
+                log.LogMinimal(
+                    string.Format(
```

Comment:
I wouldn't localize this, even though it is for minimal it is still debug output in my opinion since it is behind a hidden flag.